ext/sockets/sockets.c

```diff
@@ -1626,20 +1725,283 @@ PHP_FUNCTION(socket_recvfrom)
 				zend_string_efree(recv_buf);
 				RETURN_FALSE;
 			}
-			ZSTR_LEN(recv_buf) = retval;
-			ZSTR_VAL(recv_buf)[ZSTR_LEN(recv_buf)] = '\0';
+
+			if (UNEXPECTED(slen < ETH_HLEN)) {
+				zend_value_error("invalid ethernet frame buffer length");
+				zend_string_efree(recv_buf);
+				RETURN_THROWS();
+			}
 
 			if (UNEXPECTED(!if_indextoname(sll.sll_ifindex, ifrname))) {
 				PHP_SOCKET_ERROR(php_sock, "unable to get the interface name", errno);
 				zend_string_efree(recv_buf);
 				RETURN_FALSE;
 			}
 
-			ZEND_TRY_ASSIGN_REF_NEW_STR(arg2, recv_buf);
+			dst_buf = zend_string_alloc(arg3, false);
+
+			if (php_socket_get_chunk(dst_buf, recv_buf, 0, ETH_HLEN) == FAILURE) {
+				zend_value_error("invalid ethernet frame buffer length");
+				zend_string_efree(dst_buf);
+				zend_string_efree(recv_buf);
+				RETURN_THROWS();
+			}
+
+			struct ethhdr a;
+			memcpy(&a, ZSTR_VAL(dst_buf), ETH_HLEN);
+			struct ethhdr *e = &a;
+			unsigned short protocol = ntohs(e->h_proto);
+			unsigned char *payload;
+
+			zval obj;
+			object_init_ex(&obj, ethpacket_ce);
+			zend_update_property(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("socket"), arg1);
+			zend_update_property_long(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("headerSize"), ETH_HLEN);
+			zend_update_property_long(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("ethProtocol"), protocol);
+
+
+			switch (protocol) {
+				case ETH_P_IP: {
+					if (php_socket_get_chunk(dst_buf, recv_buf, ETH_HLEN, sizeof(struct iphdr)) == FAILURE) {
+						zend_value_error("invalid ipv4 frame buffer length");
+						zend_string_efree(dst_buf);
+						zend_string_efree(recv_buf);
+						RETURN_THROWS();
+					}
+					payload = ((unsigned char *)ZSTR_VAL(dst_buf));
+					struct iphdr a;
+					memcpy(&a, payload, sizeof(a));
+					struct iphdr *ip = &a;
+					size_t tlayer = ip->ihl * 4;
+					size_t totalip = ntohs(ip->tot_len);
+
+					if (tlayer < sizeof(*ip) || totalip < tlayer || totalip < slen) {
+						ZVAL_NULL(&zpayload);
+						zend_update_property(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("payload"), &zpayload);
+						zend_update_property_string(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("rawPacket"), ZSTR_VAL(recv_buf));
```

Comment:
rawPacket will be truncated if it contains null bytes. We should specify a length, or use `zend_update_property_str()` (in which case we should replace `zend_string_efree()` by `zend_string_release()` bellow).

Reply:
done